# Visualizing synteny and homology of genes

This repository contains a tutorial and scripts to create a gene synteny and homology figure for a set of species. 

# Required software

If you have access to the draco high-performance cluster and are part of the VEO group, you only need to install gggenomes and inkscape locally in your computer. The other software are already installed in draco and can be used as mentioned in this tutorial. In other cases, you should install the following software according to the developer's recommendations:

- [gggenomes](https://github.com/thackl/gggenomes)
- [Inkscape](https://wiki.inkscape.org/wiki/Installing_Inkscape)
- [Prodigal](https://github.com/hyattpd/Prodigal) 
- [BLAST](https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/)
- [python3](https://www.python.org/downloads/) 
- [Jupyter notebook](https://jupyter.org/install)
- [Barrnap](https://github.com/tseemann/barrnap)
- [BEDtools](https://bedtools.readthedocs.io/en/latest/index.html)
- [MAFFT](https://mafft.cbrc.jp/alignment/software/installation_without_root.html)
- [iqtree](http://www.iqtree.org/)

