{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8bf3368b-d894-4b9d-8a50-c289395bcd77",
   "metadata": {},
   "source": [
    "# Get objects for gggenomes plot\n",
    "\n",
    "This notebook creates objects required by gggenomes to do a gene synteny and homology figure. More details on https://github.com/MGXlab/genes_synteny."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c7354b19-d2b3-47e0-b3c3-6d58851fca7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Activate print all activities\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0609aa2-d6a0-403d-81a9-184a89638368",
   "metadata": {},
   "source": [
    "# Open alv_genes_blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fa895640-a56d-492a-a636-61f5edb4a9b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feat_id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DD5b</td>\n",
       "      <td>46</td>\n",
       "      <td>373085</td>\n",
       "      <td>373903</td>\n",
       "      <td>+</td>\n",
       "      <td>sumT_DD5b</td>\n",
       "      <td>sumT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>3717908</td>\n",
       "      <td>3719089</td>\n",
       "      <td>+</td>\n",
       "      <td>moeA_1_DSM1208</td>\n",
       "      <td>moeA_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>3719535</td>\n",
       "      <td>3719975</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_1_DSM1208</td>\n",
       "      <td>moaE_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>4313511</td>\n",
       "      <td>4313753</td>\n",
       "      <td>+</td>\n",
       "      <td>moaD_DSM1208</td>\n",
       "      <td>moaD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>4313763</td>\n",
       "      <td>4314173</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_2_DSM1208</td>\n",
       "      <td>moaE_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    bin_id  seq_id    start      end strand         feat_id    name\n",
       "0     DD5b      46   373085   373903      +       sumT_DD5b    sumT\n",
       "1  DSM1208       1  3717908  3719089      +  moeA_1_DSM1208  moeA_1\n",
       "2  DSM1208       1  3719535  3719975      +  moaE_1_DSM1208  moaE_1\n",
       "3  DSM1208       1  4313511  4313753      +    moaD_DSM1208    moaD\n",
       "4  DSM1208       1  4313763  4314173      +  moaE_2_DSM1208  moaE_2"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "alv_genes = pd.read_csv('../objects/alv_genes1.csv', header = 0) \n",
    "alv_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6aba4898-0e08-41e1-8db5-8c1ea3d0ddae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feat_id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DD5b</td>\n",
       "      <td>46</td>\n",
       "      <td>373085</td>\n",
       "      <td>373903</td>\n",
       "      <td>+</td>\n",
       "      <td>sumT_DD5b</td>\n",
       "      <td>sumT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>3717908</td>\n",
       "      <td>3719089</td>\n",
       "      <td>+</td>\n",
       "      <td>moeA_DSM1208</td>\n",
       "      <td>moeA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>3719535</td>\n",
       "      <td>3719975</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_DSM1208</td>\n",
       "      <td>moaE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>4313511</td>\n",
       "      <td>4313753</td>\n",
       "      <td>+</td>\n",
       "      <td>moaD_DSM1208</td>\n",
       "      <td>moaD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>4313763</td>\n",
       "      <td>4314173</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_DSM1208</td>\n",
       "      <td>moaE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    bin_id  seq_id    start      end strand       feat_id  name\n",
       "0     DD5b      46   373085   373903      +     sumT_DD5b  sumT\n",
       "1  DSM1208       1  3717908  3719089      +  moeA_DSM1208  moeA\n",
       "2  DSM1208       1  3719535  3719975      +  moaE_DSM1208  moaE\n",
       "3  DSM1208       1  4313511  4313753      +  moaD_DSM1208  moaD\n",
       "4  DSM1208       1  4313763  4314173      +  moaE_DSM1208  moaE"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Remove the underscore and number of paralog of all genes\n",
    "alv_genes['feat_id'] = alv_genes['feat_id'].str.replace(r'_1', '', regex=True)\n",
    "alv_genes['feat_id'] = alv_genes['feat_id'].str.replace(r'_2', '', regex=True)\n",
    "\n",
    "alv_genes['name'] = alv_genes['name'].str.replace(r'_1', '', regex=True)\n",
    "alv_genes['name'] = alv_genes['name'].str.replace(r'_2', '', regex=True)\n",
    "\n",
    "#Alvaro said moaC and moaC2 are NOT the same!\n",
    "#alv_genes['feat_id'] = alv_genes['feat_id'].str.replace(r'moaC2', 'moaC', regex=True)\n",
    "#alv_genes['name'] = alv_genes['name'].str.replace(r'moaC2', 'moaC', regex=True)\n",
    "alv_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7d05d5d3-720e-4fad-bb9a-4344063b540b",
   "metadata": {},
   "outputs": [],
   "source": [
    "alv_genes.to_csv('../objects/alv_genes.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc038b8a-e5e4-4fda-bdd1-b14bd259d4d3",
   "metadata": {},
   "source": [
    "# Create alv_seqs_blast\n",
    "\n",
    "Create object alv_seqs based on information of: https://thackl.github.io/gggenomes/articles/gggenomes.html#sequence-track---there-can-be-only-one\n",
    "\n",
    "Columns: seq_id,bin_id,length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9cfcf7f2-47f5-4f09-8ccf-07ba54d94524",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ZONMW-30', 'ZONMW-20', 'HI1', 'DSM1208', 'DSM15718', 'DD5b', 'UW101', 'IR1']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Put list in order of phylogenetic tree\n",
    "sp_lst = ['ZONMW-30', 'ZONMW-20', 'HI1', 'DSM1208', 'DSM15718', 'DD5b', 'UW101', 'IR1']\n",
    "sp_lst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "139a5474-e8d1-4fb3-8aab-c0ed993bac2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create an empty DataFrame with specified column names\n",
    "alv_seqs = pd.DataFrame(columns=[\"bin_id\", \"seq_id\", \"length\"])\n",
    "\n",
    "empty = 0\n",
    "\n",
    "# Iterate through the list (each species is already in the correct order) \n",
    "for bin_id in sp_lst:\n",
    "    \n",
    "    #Get ortholog genes for this genome ID\n",
    "    tmp = alv_genes[alv_genes['bin_id'] == bin_id]\n",
    "    \n",
    "    #Get list of seq_ids (contigs)\n",
    "    seq_lst = tmp['seq_id'].unique().tolist()\n",
    "    \n",
    "    #If tmp is empty, it means the name correspondence is not correct\n",
    "    if (tmp.empty):\n",
    "        print(\"Error found in looking for orthologs in alv_genes_blast!\")\n",
    "    else:\n",
    "\n",
    "        #Get min and max lengths for every bin_id (contig)\n",
    "        for seq in seq_lst:\n",
    "\n",
    "            #Get ortholog genes for this bin_id\n",
    "            tmp2 = tmp[tmp['seq_id'] == seq]\n",
    "        \n",
    "            #Get min and max reported lengths for each bin_id of this seq_id \n",
    "            max = tmp2[['start','end']].max().max()\n",
    "            min = tmp2[['start','end']].min().min()\n",
    "        \n",
    "            width = max - min + 1\n",
    "            #I will set the length of the contig as the max position plus ten\n",
    "            length = int(max + 10)\n",
    "\n",
    "            # Append a new row to the DataFrame with the values\n",
    "            new_row = pd.Series({\"bin_id\": bin_id, \"seq_id\": str(seq), \"length\": length})\n",
    "            alv_seqs = pd.concat([alv_seqs, pd.DataFrame([new_row])], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "398801a4-b2e9-4b92-a43b-2899cddcb418",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>172314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>7</td>\n",
       "      <td>11335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>13</td>\n",
       "      <td>19644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>10</td>\n",
       "      <td>176073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>17</td>\n",
       "      <td>121851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HI1</td>\n",
       "      <td>1</td>\n",
       "      <td>3188588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>4337566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DSM15718</td>\n",
       "      <td>30</td>\n",
       "      <td>51564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DD5b</td>\n",
       "      <td>46</td>\n",
       "      <td>373913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>UW101</td>\n",
       "      <td>1</td>\n",
       "      <td>5613620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>IR1</td>\n",
       "      <td>1</td>\n",
       "      <td>415594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>IR1</td>\n",
       "      <td>856</td>\n",
       "      <td>32648</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      bin_id seq_id   length\n",
       "0   ZONMW-30      4   172314\n",
       "1   ZONMW-30      7    11335\n",
       "2   ZONMW-30     13    19644\n",
       "3   ZONMW-20     10   176073\n",
       "4   ZONMW-20     17   121851\n",
       "5        HI1      1  3188588\n",
       "6    DSM1208      1  4337566\n",
       "7   DSM15718     30    51564\n",
       "8       DD5b     46   373913\n",
       "9      UW101      1  5613620\n",
       "10       IR1      1   415594\n",
       "11       IR1    856    32648"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>172314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>7</td>\n",
       "      <td>11335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>13</td>\n",
       "      <td>19644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>10</td>\n",
       "      <td>176073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>17</td>\n",
       "      <td>121851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HI1</td>\n",
       "      <td>1</td>\n",
       "      <td>3188588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>4337566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DSM15718</td>\n",
       "      <td>30</td>\n",
       "      <td>51564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DD5b</td>\n",
       "      <td>46</td>\n",
       "      <td>373913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>UW101</td>\n",
       "      <td>1</td>\n",
       "      <td>5613620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>IR1</td>\n",
       "      <td>856</td>\n",
       "      <td>32648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>IR1</td>\n",
       "      <td>1</td>\n",
       "      <td>415594</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      bin_id seq_id   length\n",
       "0   ZONMW-30      4   172314\n",
       "1   ZONMW-30      7    11335\n",
       "2   ZONMW-30     13    19644\n",
       "3   ZONMW-20     10   176073\n",
       "4   ZONMW-20     17   121851\n",
       "5        HI1      1  3188588\n",
       "6    DSM1208      1  4337566\n",
       "7   DSM15718     30    51564\n",
       "8       DD5b     46   373913\n",
       "9      UW101      1  5613620\n",
       "10       IR1    856    32648\n",
       "11       IR1      1   415594"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alv_seqs\n",
    "\n",
    "#Swap rows of indexes 2 and row 3 - To change the order of contigs of species IR1, put the longest first\n",
    "alv_seqs.iloc[[10, 11]] = alv_seqs.iloc[[11, 10]].values\n",
    "\n",
    "alv_seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6c1cb7e5-a784-4019-85e0-c37be32013e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "alv_seqs.to_csv('../objects/alv_seqs.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b64597cd-06f7-455a-93b3-75b867a96ba2",
   "metadata": {},
   "source": [
    "# Create alv_seqs_blast2 and alv_genes_blast2 with shortened genome length\n",
    "\n",
    "To improve visualization, the length of the genomes in alv_seqs and the start and end of genes in alv_genes is shortened.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2ca83f44-329a-4839-ae4b-ecf8d6095ea5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feat_id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DD5b</td>\n",
       "      <td>46</td>\n",
       "      <td>373085</td>\n",
       "      <td>373903</td>\n",
       "      <td>+</td>\n",
       "      <td>sumT_DD5b</td>\n",
       "      <td>sumT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>3717908</td>\n",
       "      <td>3719089</td>\n",
       "      <td>+</td>\n",
       "      <td>moeA_DSM1208</td>\n",
       "      <td>moeA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>3719535</td>\n",
       "      <td>3719975</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_DSM1208</td>\n",
       "      <td>moaE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>4313511</td>\n",
       "      <td>4313753</td>\n",
       "      <td>+</td>\n",
       "      <td>moaD_DSM1208</td>\n",
       "      <td>moaD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>4313763</td>\n",
       "      <td>4314173</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_DSM1208</td>\n",
       "      <td>moaE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    bin_id  seq_id    start      end strand       feat_id  name\n",
       "0     DD5b      46   373085   373903      +     sumT_DD5b  sumT\n",
       "1  DSM1208       1  3717908  3719089      +  moeA_DSM1208  moeA\n",
       "2  DSM1208       1  3719535  3719975      +  moaE_DSM1208  moaE\n",
       "3  DSM1208       1  4313511  4313753      +  moaD_DSM1208  moaD\n",
       "4  DSM1208       1  4313763  4314173      +  moaE_DSM1208  moaE"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>172314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>7</td>\n",
       "      <td>11335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>13</td>\n",
       "      <td>19644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>10</td>\n",
       "      <td>176073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>17</td>\n",
       "      <td>121851</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     bin_id  seq_id  length\n",
       "0  ZONMW-30       4  172314\n",
       "1  ZONMW-30       7   11335\n",
       "2  ZONMW-30      13   19644\n",
       "3  ZONMW-20      10  176073\n",
       "4  ZONMW-20      17  121851"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "alv_genes = pd.read_csv('../objects/alv_genes.csv', header = 0) #Not anchored\n",
    "alv_seqs = pd.read_csv('../objects/alv_seqs.csv', header = 0, index_col=False)\n",
    "alv_genes.head()\n",
    "alv_seqs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "92cf37b2-eb6f-4842-a142-ae5e96dcbde1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ZONMW-30', 'ZONMW-20', 'HI1', 'DSM1208', 'DSM15718', 'DD5b', 'UW101', 'IR1']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Get list of species in the right order\n",
    "sp_lst = list(dict.fromkeys(alv_seqs['bin_id'] ))\n",
    "sp_lst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d1c4aad8-1c98-422f-9905-f30f3a30516e",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### ON USE ######\n",
    "\n",
    "###################\n",
    "\n",
    "#This code subtracts starts and ends of genomes\n",
    "# based on global starts and ends\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "#Copy dataframes\n",
    "alv_genes2 = alv_genes.copy()\n",
    "alv_seqs2 = alv_seqs.copy()\n",
    "\n",
    "#Go through every species (bin_id)\n",
    "for sp in sp_lst:\n",
    "    \n",
    "    #Select all genes of the current species\n",
    "    alv_genes_bin = alv_genes[alv_genes['bin_id'] == sp]\n",
    "    \n",
    "    #Go through every seq_id (contig)\n",
    "    seq_lst = alv_genes_bin['seq_id'].unique().tolist()\n",
    "    \n",
    "    #Go through every contig (seq_id)\n",
    "    for seq_id in seq_lst:\n",
    "\n",
    "        #Select all genes of the current species(bin_id) and seq_id\n",
    "        alv_genes_seq = alv_genes_bin[alv_genes_bin['seq_id'] == seq_id]        \n",
    "        \n",
    "        #Get min reported positions of all genes\n",
    "        min = alv_genes_seq[['start','end']].min().min()\n",
    "        \n",
    "        ### Substract min length from sequences' start and end ###\n",
    "        alv_genes2.loc[(alv_genes2['bin_id'] == sp) & (alv_genes2['seq_id'] == seq_id), 'start' ] -=  min\n",
    "        alv_genes2.loc[(alv_genes2['bin_id'] == sp) & (alv_genes2['seq_id'] == seq_id), 'end' ] -=  min\n",
    "        \n",
    "        ### Substract min length from genomes ###\n",
    "        alv_seqs2.loc[(alv_seqs2['bin_id'] == sp) & (alv_seqs2['seq_id'] == seq_id), 'length' ] -=  min - 1       \n",
    "        \n",
    "#alv_genes2['start'] = alv_genes2['start'].astype(int)\n",
    "#alv_genes2['end'] = alv_genes2['end'].astype(int)\n",
    "#alv_seqs2['length'] = alv_seqs2['start'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "00d34835-5ec0-4d2e-9721-330bbe3dbd3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feat_id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DD5b</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>818</td>\n",
       "      <td>+</td>\n",
       "      <td>sumT_DD5b</td>\n",
       "      <td>sumT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1181</td>\n",
       "      <td>+</td>\n",
       "      <td>moeA_DSM1208</td>\n",
       "      <td>moeA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>1627</td>\n",
       "      <td>2067</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_DSM1208</td>\n",
       "      <td>moaE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>595603</td>\n",
       "      <td>595845</td>\n",
       "      <td>+</td>\n",
       "      <td>moaD_DSM1208</td>\n",
       "      <td>moaD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>595855</td>\n",
       "      <td>596265</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_DSM1208</td>\n",
       "      <td>moaE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>596428</td>\n",
       "      <td>597327</td>\n",
       "      <td>+</td>\n",
       "      <td>moaA_DSM1208</td>\n",
       "      <td>moaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>597332</td>\n",
       "      <td>597805</td>\n",
       "      <td>+</td>\n",
       "      <td>moaC_DSM1208</td>\n",
       "      <td>moaC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>597789</td>\n",
       "      <td>598385</td>\n",
       "      <td>+</td>\n",
       "      <td>mobA_DSM1208</td>\n",
       "      <td>mobA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>598382</td>\n",
       "      <td>599455</td>\n",
       "      <td>+</td>\n",
       "      <td>moeZ_DSM1208</td>\n",
       "      <td>moeZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>618458</td>\n",
       "      <td>619648</td>\n",
       "      <td>-</td>\n",
       "      <td>moeA_DSM1208</td>\n",
       "      <td>moeA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    bin_id  seq_id   start     end strand       feat_id  name\n",
       "0     DD5b      46       0     818      +     sumT_DD5b  sumT\n",
       "1  DSM1208       1       0    1181      +  moeA_DSM1208  moeA\n",
       "2  DSM1208       1    1627    2067      +  moaE_DSM1208  moaE\n",
       "3  DSM1208       1  595603  595845      +  moaD_DSM1208  moaD\n",
       "4  DSM1208       1  595855  596265      +  moaE_DSM1208  moaE\n",
       "5  DSM1208       1  596428  597327      +  moaA_DSM1208  moaA\n",
       "6  DSM1208       1  597332  597805      +  moaC_DSM1208  moaC\n",
       "7  DSM1208       1  597789  598385      +  mobA_DSM1208  mobA\n",
       "8  DSM1208       1  598382  599455      +  moeZ_DSM1208  moeZ\n",
       "9  DSM1208       1  618458  619648      -  moeA_DSM1208  moeA"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>154620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>7</td>\n",
       "      <td>2428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>13</td>\n",
       "      <td>1690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>10</td>\n",
       "      <td>77708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>17</td>\n",
       "      <td>595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HI1</td>\n",
       "      <td>1</td>\n",
       "      <td>2392577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>619659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DSM15718</td>\n",
       "      <td>30</td>\n",
       "      <td>805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DD5b</td>\n",
       "      <td>46</td>\n",
       "      <td>829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>UW101</td>\n",
       "      <td>1</td>\n",
       "      <td>7527</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     bin_id  seq_id   length\n",
       "0  ZONMW-30       4   154620\n",
       "1  ZONMW-30       7     2428\n",
       "2  ZONMW-30      13     1690\n",
       "3  ZONMW-20      10    77708\n",
       "4  ZONMW-20      17      595\n",
       "5       HI1       1  2392577\n",
       "6   DSM1208       1   619659\n",
       "7  DSM15718      30      805\n",
       "8      DD5b      46      829\n",
       "9     UW101       1     7527"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alv_genes2.head(n=10)\n",
    "alv_seqs2.head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5f03e7d2-3b5c-426a-8445-0a0b2c92bef1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "52"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(alv_genes2[alv_genes2['start'] >= 0])\n",
    "len(alv_genes2[alv_genes2['start'] < 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6c516144-3add-4909-bdf7-5ce8e551522f",
   "metadata": {},
   "outputs": [],
   "source": [
    "alv_seqs2.to_csv('../objects/alv_seqs2.csv', index = False)\n",
    "alv_genes2.to_csv('../objects/alv_genes2.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf995c4f-76d3-44c0-b4b5-6c8f97917db8",
   "metadata": {},
   "source": [
    "# Create alv_ava_blast2\n",
    "\n",
    "alv_ava contains the homology between contigs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cd342566-cd32-4165-b0e4-64a069929153",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>172314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>7</td>\n",
       "      <td>11335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>13</td>\n",
       "      <td>19644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>10</td>\n",
       "      <td>176073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>17</td>\n",
       "      <td>121851</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     bin_id  seq_id  length\n",
       "0  ZONMW-30       4  172314\n",
       "1  ZONMW-30       7   11335\n",
       "2  ZONMW-30      13   19644\n",
       "3  ZONMW-20      10  176073\n",
       "4  ZONMW-20      17  121851"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "alv_seqs2 = pd.read_csv('../objects/alv_seqs2.csv', header = 0, index_col=False)\n",
    "alv_seqs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "840424c3-8d2e-4f25-a410-fddcce299ef4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ZONMW-30', 'ZONMW-20', 'HI1', 'DSM1208', 'DSM15718', 'DD5b', 'UW101', 'IR1']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>bin_id2</th>\n",
       "      <th>seq_id2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>7</td>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>7</td>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>13</td>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     bin_id seq_id   bin_id2 seq_id2\n",
       "0  ZONMW-30      4  ZONMW-20      10\n",
       "1  ZONMW-30      4  ZONMW-20      17\n",
       "2  ZONMW-30      7  ZONMW-20      10\n",
       "3  ZONMW-30      7  ZONMW-20      17\n",
       "4  ZONMW-30     13  ZONMW-20      10"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#New ON USE###\n",
    "##################\n",
    "\n",
    "#Create empty DataFrame with specified column names\n",
    "alv_ava2 = pd.DataFrame(columns=[\"bin_id\", \"seq_id\", \"bin_id2\", \"seq_id2\"])\n",
    "\n",
    "#Get list of species in the right order\n",
    "sp_lst = list(dict.fromkeys(alv_seqs['bin_id'] ))\n",
    "sp_lst\n",
    "\n",
    "#For every line of alv_species, assign genome homology/orthology\n",
    "for i in range(len(sp_lst)):\n",
    "    \n",
    "    #Do not make comparisons with the last species\n",
    "    if(i < 7):\n",
    "\n",
    "        bin_id = sp_lst[i]\n",
    "        bin_id2 = sp_lst[i + 1]\n",
    "\n",
    "        #Select all seqs (contigs) of the current bin/species\n",
    "        alv_seqs2_bin = alv_seqs2[alv_seqs2['bin_id'] == bin_id]\n",
    "\n",
    "        #Select all seqs (contigs) of the next bin/species\n",
    "        alv_seqs2_bin2 = alv_seqs2[alv_seqs2['bin_id'] == bin_id2]\n",
    "        \n",
    "        #Get list of every seq_id (contig) of the current bin/species\n",
    "        seq_lst = alv_seqs2_bin['seq_id'].unique().tolist()\n",
    "\n",
    "        #Get list of every seq_id (contig) of the next bin/species\n",
    "        seq_lst2 = alv_seqs2_bin2['seq_id'].unique().tolist()\n",
    "    \n",
    "        #Go through every contig (seq_id)\n",
    "        for seq_id in seq_lst:\n",
    "\n",
    "            #Go through every contig (seq_id)\n",
    "            for seq_id2 in seq_lst2:\n",
    "            \n",
    "                # Append a new row to the DataFrame with the values\n",
    "                new_row = pd.Series({\"bin_id\": bin_id, \"seq_id\": seq_id, \"bin_id2\": bin_id2, \"seq_id2\": seq_id2})\n",
    "                alv_ava2 = pd.concat([alv_ava2, pd.DataFrame([new_row])], ignore_index=True)\n",
    "\n",
    "alv_ava2.head()        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cc85e1e4-9b51-45f2-9b91-075b0c432370",
   "metadata": {},
   "outputs": [],
   "source": [
    "alv_ava2.to_csv('../objects/alv_ava2.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4b8df4f-b689-4ab8-bba8-27519e77c959",
   "metadata": {},
   "source": [
    "# Create alv_prot_ava_blast2\n",
    "\n",
    "alv_ava contains the homology between genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c39bdb7c-7f26-4a1a-ad50-2081d21a5d4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feat_id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DD5b</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>818</td>\n",
       "      <td>+</td>\n",
       "      <td>sumT_DD5b</td>\n",
       "      <td>sumT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1181</td>\n",
       "      <td>+</td>\n",
       "      <td>moeA_DSM1208</td>\n",
       "      <td>moeA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>1627</td>\n",
       "      <td>2067</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_DSM1208</td>\n",
       "      <td>moaE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>595603</td>\n",
       "      <td>595845</td>\n",
       "      <td>+</td>\n",
       "      <td>moaD_DSM1208</td>\n",
       "      <td>moaD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>595855</td>\n",
       "      <td>596265</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_DSM1208</td>\n",
       "      <td>moaE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    bin_id  seq_id   start     end strand       feat_id  name\n",
       "0     DD5b      46       0     818      +     sumT_DD5b  sumT\n",
       "1  DSM1208       1       0    1181      +  moeA_DSM1208  moeA\n",
       "2  DSM1208       1    1627    2067      +  moaE_DSM1208  moaE\n",
       "3  DSM1208       1  595603  595845      +  moaD_DSM1208  moaD\n",
       "4  DSM1208       1  595855  596265      +  moaE_DSM1208  moaE"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "52"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "alv_genes2.to_csv('../objects/alv_genes2.csv', index = False)      \n",
    "alv_genes2.head()\n",
    "len(alv_genes2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "81cd57bd-c63f-4307-95cf-f325b7bc377a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>feat_id</th>\n",
       "      <th>bin_id2</th>\n",
       "      <th>feat_id2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>moaE_ZONMW-30</td>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>moaE_ZONMW-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>moaA_ZONMW-30</td>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>moaA_ZONMW-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>moeA_ZONMW-30</td>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>moeA_ZONMW-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>mobA_ZONMW-30</td>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>mobA_ZONMW-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>moeA_ZONMW-30</td>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>moeA_ZONMW-20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     bin_id        feat_id   bin_id2       feat_id2\n",
       "0  ZONMW-30  moaE_ZONMW-30  ZONMW-20  moaE_ZONMW-20\n",
       "1  ZONMW-30  moaA_ZONMW-30  ZONMW-20  moaA_ZONMW-20\n",
       "2  ZONMW-30  moeA_ZONMW-30  ZONMW-20  moeA_ZONMW-20\n",
       "3  ZONMW-30  mobA_ZONMW-30  ZONMW-20  mobA_ZONMW-20\n",
       "4  ZONMW-30  moeA_ZONMW-30  ZONMW-20  moeA_ZONMW-20"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Create empty DataFrame with specified column names\n",
    "alv_prot_ava2 = pd.DataFrame(columns=[\"bin_id\", \"feat_id\",\n",
    "                                      \"bin_id2\", \"feat_id2\"])\n",
    "\n",
    "pd.set_option('mode.chained_assignment', None)\n",
    "import numpy as np\n",
    "\n",
    "#Get list of species in the right order\n",
    "sp_lst = list(dict.fromkeys(alv_seqs['bin_id'] ))\n",
    "\n",
    "#For every line of alv_species, get genes and assign homologies\n",
    "for i in range(len(sp_lst)):\n",
    "    \n",
    "    #Get species and next species\n",
    "    sp = sp_lst[i]\n",
    "\n",
    "    #print(i, len(sp_lst))\n",
    "    \n",
    "    #Do not make comparisons with the last species\n",
    "    if(i < (len(sp_lst) - 1) ):    \n",
    "\n",
    "        sp2 = sp_lst[i + 1]\n",
    "        #print(sp, sp2)\n",
    "        \n",
    "        #Select all genes of the current species\n",
    "        alv_genes_bin = alv_genes[alv_genes['bin_id'] == sp]\n",
    "\n",
    "        #Select all genes of the next species\n",
    "        alv_genes_bin2 = alv_genes[alv_genes['bin_id'] == sp2]\n",
    "        \n",
    "        #Go through every gene of current species and compare to all genes of next species. Assign homologies \n",
    "        for x in range (len(alv_genes_bin)):\n",
    "        \n",
    "            #Get information for current gene\n",
    "            row1 = alv_genes_bin.iloc[x]\n",
    "            feat_id1 = row1.loc['feat_id']\n",
    "            name1 = row1.loc['name']\n",
    "\n",
    "            for x2 in range (len(alv_genes_bin2)):\n",
    "            \n",
    "                #Get information for current element\n",
    "                row2 = alv_genes_bin2.iloc[x2]\n",
    "                feat_id2 = row2.loc['feat_id']\n",
    "                name2 = row2.loc['name']\n",
    "\n",
    "                #Found homology!\n",
    "                if(name1 == name2):\n",
    "                    \n",
    "                    #Add homology to file\n",
    "                    homology_row = pd.Series({\"bin_id\": sp, \"feat_id\": feat_id1, \n",
    "                                              \"bin_id2\": sp2, \"feat_id2\": feat_id2})\n",
    "                    \n",
    "                    alv_prot_ava2 = pd.concat([alv_prot_ava2, pd.DataFrame([homology_row])], ignore_index=True) # spacer\n",
    "alv_prot_ava2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9c8d9fb1-e5e8-49a9-bf73-9a268899c62e",
   "metadata": {},
   "outputs": [],
   "source": [
    "alv_prot_ava2 = alv_prot_ava2.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c5bcd435-a4eb-4da6-8bc5-5428204396a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "alv_prot_ava2.to_csv('../objects/alv_prot_ava2.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5af8c056-a152-4b4e-9faf-c02afc51ef35",
   "metadata": {},
   "source": [
    "# Decrease distances between elements\n",
    "\n",
    "Create: alv_ava_blast3 and alv_genes_blast3.  \n",
    "\n",
    "To improve visualization of the synteny, all large spaces between elements (genes or operons) are to be substituted by a spacer element of X bp surrounded by Y bp.   \n",
    "\n",
    "The only object to change is alv_all2.csv, all others are to be kept the same."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "af6442d8-d66f-47e2-ae74-2d3ff3ef99f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Activate print all activities\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ce505b22-4986-4ad2-b186-59b9518b92b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feat_id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DD5b</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>818</td>\n",
       "      <td>+</td>\n",
       "      <td>sumT_DD5b</td>\n",
       "      <td>sumT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1181</td>\n",
       "      <td>+</td>\n",
       "      <td>moeA_DSM1208</td>\n",
       "      <td>moeA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>1627</td>\n",
       "      <td>2067</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_DSM1208</td>\n",
       "      <td>moaE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>595603</td>\n",
       "      <td>595845</td>\n",
       "      <td>+</td>\n",
       "      <td>moaD_DSM1208</td>\n",
       "      <td>moaD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSM1208</td>\n",
       "      <td>1</td>\n",
       "      <td>595855</td>\n",
       "      <td>596265</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_DSM1208</td>\n",
       "      <td>moaE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    bin_id  seq_id   start     end strand       feat_id  name\n",
       "0     DD5b      46       0     818      +     sumT_DD5b  sumT\n",
       "1  DSM1208       1       0    1181      +  moeA_DSM1208  moeA\n",
       "2  DSM1208       1    1627    2067      +  moaE_DSM1208  moaE\n",
       "3  DSM1208       1  595603  595845      +  moaD_DSM1208  moaD\n",
       "4  DSM1208       1  595855  596265      +  moaE_DSM1208  moaE"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>154620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>7</td>\n",
       "      <td>2428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>13</td>\n",
       "      <td>1690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>10</td>\n",
       "      <td>77708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>17</td>\n",
       "      <td>595</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     bin_id  seq_id  length\n",
       "0  ZONMW-30       4  154620\n",
       "1  ZONMW-30       7    2428\n",
       "2  ZONMW-30      13    1690\n",
       "3  ZONMW-20      10   77708\n",
       "4  ZONMW-20      17     595"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "alv_genes2 = pd.read_csv('../objects/alv_genes2.csv', header = 0)\n",
    "alv_seqs2 = pd.read_csv('../objects/alv_seqs2.csv', header = 0, index_col=False)\n",
    "alv_genes2.head()\n",
    "alv_seqs2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b5043d90-69c4-424f-a6cb-b6075d8c82b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feat_id</th>\n",
       "      <th>name</th>\n",
       "      <th>width</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td>START</td>\n",
       "      <td>START</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>443</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_ZONMW-30</td>\n",
       "      <td>moaE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>443</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_ZONMW-30</td>\n",
       "      <td>moaE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>444</td>\n",
       "      <td>1340</td>\n",
       "      <td>+</td>\n",
       "      <td>moaC2_ZONMW-30</td>\n",
       "      <td>moaC2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>444</td>\n",
       "      <td>1340</td>\n",
       "      <td>+</td>\n",
       "      <td>moaC2_ZONMW-30</td>\n",
       "      <td>moaC2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     bin_id seq_id start   end strand         feat_id   name  width\n",
       "0  ZONMW-30      4     1   100      +           START  START    NaN\n",
       "1  ZONMW-30      4     0   443      +   moaE_ZONMW-30   moaE    NaN\n",
       "2  ZONMW-30      4     0   443      +   moaE_ZONMW-30   moaE    NaN\n",
       "3  ZONMW-30      4   444  1340      +  moaC2_ZONMW-30  moaC2    NaN\n",
       "4  ZONMW-30      4   444  1340      +  moaC2_ZONMW-30  moaC2    NaN"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feat_id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>443</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_ZONMW-30</td>\n",
       "      <td>moaE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>444</td>\n",
       "      <td>1340</td>\n",
       "      <td>+</td>\n",
       "      <td>moaC2_ZONMW-30</td>\n",
       "      <td>moaC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>1333</td>\n",
       "      <td>2325</td>\n",
       "      <td>+</td>\n",
       "      <td>moaA_ZONMW-30</td>\n",
       "      <td>moaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>2325</td>\n",
       "      <td>3491</td>\n",
       "      <td>+</td>\n",
       "      <td>moeA_ZONMW-30</td>\n",
       "      <td>moeA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>3991</td>\n",
       "      <td>4242</td>\n",
       "      <td>+</td>\n",
       "      <td>z_spacer</td>\n",
       "      <td>z_spacer</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     bin_id seq_id start   end strand         feat_id      name\n",
       "1  ZONMW-30      4     0   443      +   moaE_ZONMW-30      moaE\n",
       "3  ZONMW-30      4   444  1340      +  moaC2_ZONMW-30     moaC2\n",
       "5  ZONMW-30      4  1333  2325      +   moaA_ZONMW-30      moaA\n",
       "7  ZONMW-30      4  2325  3491      +   moeA_ZONMW-30      moeA\n",
       "8  ZONMW-30      4  3991  4242      +        z_spacer  z_spacer"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('mode.chained_assignment', None)\n",
    "import numpy as np\n",
    "\n",
    "#Numbers below add a really long spacer between genes\n",
    "#max_dist = 10000\n",
    "#spacer = 2500\n",
    "#spacer_width = 1000\n",
    "#Numbers below offer a better visualization, with almost no spacers between genes\n",
    "max_dist = 5000\n",
    "spacer = 500\n",
    "spacer_width = 250\n",
    "\n",
    "alv_genes3 = pd.DataFrame(columns=alv_genes2.columns)\n",
    "\n",
    "#Work through every species and contig in order. If elements are farther away than max_dist, \n",
    "# remove the distance and add 2*spacer + spacer_width\n",
    "for index, row in alv_seqs2.iterrows():\n",
    "\n",
    "    #Get current bin (species) and seq (contig)\n",
    "    bin_id = row['bin_id']\n",
    "    seq_id = row['seq_id']\n",
    "    \n",
    "    #Get all elements belonging to seq and bin (genes, operons)\n",
    "    el = alv_genes2[(alv_genes2['bin_id'] == bin_id) & (alv_genes2['seq_id'] == seq_id)]\n",
    "\n",
    "    #Put elements in ascending order of start position\n",
    "    el_ordered = el.sort_values(by='start', ascending=True)\n",
    "    #Add first row with fake element starting on 1bp (this is to cut the large spaces in the begining)\n",
    "    #header:\n",
    "    #bin_id\tseq_id\tstart\tend\tstrand\tfeat_id\tname\n",
    "    first_row = pd.DataFrame({\"bin_id\": [bin_id], \"seq_id\": [seq_id], \"start\": [1], \"end\": [100], \"strand\": ['+'],\n",
    "                              \"feat_id\": ['START'], \"name\": ['START']})\n",
    "    el_ordered.reset_index(drop=True, inplace=True)\n",
    "    el_ordered = pd.concat([first_row, el_ordered]).reset_index(drop=True)\n",
    "    el_ordered.reset_index(drop=True, inplace=True)    \n",
    "    \n",
    "    #Compare 1st with second line/element, 2nd with 3rd and so on\n",
    "    #For every line (gene/operon) of alv_species, re-assign start and end positions whenever necessary\n",
    "    for i in range (len(el_ordered)):\n",
    "      \n",
    "        #Do the comparison only until the last element\n",
    "        if(i < (len(el_ordered) - 1)):\n",
    "        \n",
    "            #Get information for current element\n",
    "            row1 = el_ordered.iloc[i]\n",
    "            start1 = row1['end']\n",
    "        \n",
    "            #Get information for next element\n",
    "            row2 = el_ordered.iloc[i + 1]\n",
    "            start2 = row2['start']\n",
    "            end2 = row2['end']\n",
    "            width2 = end2 - start2 + 1\n",
    "                \n",
    "            #Check how far elements are from each other (end of element 1 and start of element 2)\n",
    "            distance = (start2 - start1)\n",
    "            \n",
    "            #If too far, include spacer, otherwise keep elements as they are\n",
    "            if (distance >  max_dist):\n",
    "                \n",
    "                #Prepare spacer\n",
    "                spacer_row = row1.copy()\n",
    "                spacer_row.loc['start'] = start1 + spacer \n",
    "                spacer_row.loc['end'] = start1 + spacer + spacer_width + 1\n",
    "                spacer_row.loc['strand'] = '+'\n",
    "                spacer_row.loc['feat_id'] = 'z_spacer'\n",
    "                spacer_row.loc['width'] = spacer_width \n",
    "                spacer_row.loc['name'] = 'z_spacer'\n",
    "                #Append spacer\n",
    "                alv_genes3 = pd.concat([alv_genes3, pd.DataFrame([spacer_row])], ignore_index=True) # spacer\n",
    "                \n",
    "                #Prepare new row2\n",
    "                new_row2 = row2                \n",
    "                new_row2.loc['start'] = start1 + (2 * spacer) + spacer_width + 1\n",
    "                new_row2.loc['end'] = (start1 + (2 * spacer) + spacer_width + 1) + width2 + 1\n",
    "                #Append row2\n",
    "                alv_genes3 = pd.concat([alv_genes3, pd.DataFrame([new_row2])], ignore_index=True) # spacer\n",
    "\n",
    "                i2 = i + 1\n",
    "                #Change el_ordered for the changed row permanently\n",
    "                el_ordered.loc[i2, 'start'] = start1 + (2 * spacer) + spacer_width + 1\n",
    "                el_ordered.loc[i2, 'end'] = (start1 + (2 * spacer) + spacer_width + 1) + width2 + 1\n",
    "                \n",
    "                #Re-set starts and ends of all further elements of the dataframe (genes and operons of this species)\n",
    "                for i3 in range ((i2+1), len(el_ordered)):\n",
    "                    \n",
    "                    el_ordered.loc[i3, 'start'] = (el_ordered.at[i3, 'start']) - distance + 2 * spacer + spacer_width + 1\n",
    "                    el_ordered.loc[i3, 'end'] = (el_ordered.at[i3, 'end']) - distance + 2 * spacer + spacer_width + 1\n",
    "            \n",
    "            else:\n",
    "                alv_genes3 = pd.concat([alv_genes3, pd.DataFrame([row1])], ignore_index=True)\n",
    "                alv_genes3 = pd.concat([alv_genes3, pd.DataFrame([row2])], ignore_index=True)\n",
    "\n",
    "#Remove duplicates since the elements are printed twice (once as row1 and once as row2)\n",
    "alv_genes3.head()\n",
    "\n",
    "alv_genes3 = alv_genes3.drop_duplicates()\n",
    "alv_genes3 = alv_genes3[~alv_genes3['feat_id'].str.contains('START')]\n",
    "#Remove unnecessary column\n",
    "alv_genes3 = alv_genes3.drop(['width'], axis=1)\n",
    "\n",
    "alv_genes3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8f284d19-daae-4ecd-b576-986587ed0642",
   "metadata": {},
   "outputs": [],
   "source": [
    "alv_genes3.to_csv('../objects/alv_genes3.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f1b6e6f-e6df-443c-a06a-a1721f8c2cdd",
   "metadata": {},
   "source": [
    "# Decrease global distances of genomes in alv_seqs3\n",
    "\n",
    "Following on the code block above, I decreased the total genome length to improve visualization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "97eba957-1e44-4e81-845b-f123439e6f56",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Activate print all activities\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4619ffd5-fb7c-44c6-8190-821a34717639",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>154620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>7</td>\n",
       "      <td>2428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>13</td>\n",
       "      <td>1690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>10</td>\n",
       "      <td>77708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZONMW-20</td>\n",
       "      <td>17</td>\n",
       "      <td>595</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     bin_id  seq_id  length\n",
       "0  ZONMW-30       4  154620\n",
       "1  ZONMW-30       7    2428\n",
       "2  ZONMW-30      13    1690\n",
       "3  ZONMW-20      10   77708\n",
       "4  ZONMW-20      17     595"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin_id</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>feat_id</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>443</td>\n",
       "      <td>+</td>\n",
       "      <td>moaE_ZONMW-30</td>\n",
       "      <td>moaE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>444</td>\n",
       "      <td>1340</td>\n",
       "      <td>+</td>\n",
       "      <td>moaC2_ZONMW-30</td>\n",
       "      <td>moaC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>1333</td>\n",
       "      <td>2325</td>\n",
       "      <td>+</td>\n",
       "      <td>moaA_ZONMW-30</td>\n",
       "      <td>moaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>2325</td>\n",
       "      <td>3491</td>\n",
       "      <td>+</td>\n",
       "      <td>moeA_ZONMW-30</td>\n",
       "      <td>moeA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZONMW-30</td>\n",
       "      <td>4</td>\n",
       "      <td>3991</td>\n",
       "      <td>4242</td>\n",
       "      <td>+</td>\n",
       "      <td>z_spacer</td>\n",
       "      <td>z_spacer</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     bin_id  seq_id  start   end strand         feat_id      name\n",
       "0  ZONMW-30       4      0   443      +   moaE_ZONMW-30      moaE\n",
       "1  ZONMW-30       4    444  1340      +  moaC2_ZONMW-30     moaC2\n",
       "2  ZONMW-30       4   1333  2325      +   moaA_ZONMW-30      moaA\n",
       "3  ZONMW-30       4   2325  3491      +   moeA_ZONMW-30      moeA\n",
       "4  ZONMW-30       4   3991  4242      +        z_spacer  z_spacer"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "alv_seqs2 = pd.read_csv('../objects/alv_seqs2.csv', header = 0, index_col=False)\n",
    "alv_genes3 = pd.read_csv('../objects/alv_genes3.csv', header = 0, index_col=False)\n",
    "alv_seqs2.head()\n",
    "alv_genes3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9d1dde8d-589d-4da8-af72-7a785ff3d479",
   "metadata": {},
   "outputs": [],
   "source": [
    "alv_seqs3 = alv_seqs2.copy()\n",
    "\n",
    "for index, row in alv_seqs2.iterrows():\n",
    "\n",
    "    #Get current bin (species) and seq (contig)\n",
    "    bin_id = row['bin_id']\n",
    "    seq_id = row['seq_id']\n",
    "    \n",
    "    #Get all elements belonging to seq and bin (genes, operons)\n",
    "    el = alv_genes3[(alv_genes3['bin_id'] == bin_id) & (alv_genes3['seq_id'] == seq_id)]\n",
    "    \n",
    "    #Get maximum values \n",
    "    max = el['end'].max()\n",
    "\n",
    "    #print(seq_id,alv_seqs3.loc[index, 'length'], max)\n",
    "    \n",
    "    #Substitute values in alv_seqs3\n",
    "    alv_seqs3.loc[index, 'length'] = max + 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "545fc2a3-025a-48fe-b1b0-abd91ec0af1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "alv_seqs3.to_csv('../objects/alv_seqs3.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "771166fe-e969-440c-85d9-01b74aca30f4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
